File: backend/src/main/java/com/autocare/serviceorder/entity/ServiceOrder.java
package com.autocare.serviceorder.entity;

import com.autocare.client.entity.Client;
import com.autocare.mechanic.entity.Mechanic;
import com.autocare.vehicle.entity.Vehicle;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "service_orders")
public class ServiceOrder {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(name = "order_number", unique = true, nullable = false)
    private String orderNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "client_id", nullable = false)
    private Client client;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "vehicle_id", nullable = false)
    private Vehicle vehicle;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "mechanic_id")
    private Mechanic mechanic;

    @Column(name = "reported_problem")
    private String reportedProblem;

    private String diagnosis;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ServiceOrderStatus status;

    @Column(name = "total_amount")
    private BigDecimal totalAmount;

    @Column(name = "started_at")
    private LocalDateTime startedAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (status == null) {
            status = ServiceOrderStatus.CRIADA;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: backend/app/models/service_order.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, Enum, ForeignKey, Numeric, Sequence, String, Uuid, text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.models.client import Client
from app.models.enums import ServiceOrderStatus
from app.models.mechanic import Mechanic
from app.models.mixins import Timestamps, UUIDPrimaryKey
from app.models.vehicle import Vehicle

# Números de OS e orçamento vêm de sequences nativas do Postgres: atômicas sob concorrência.
service_order_number_seq = Sequence("service_order_number_seq", metadata=Base.metadata)


class ServiceOrder(UUIDPrimaryKey, Timestamps, Base):
    __tablename__ = "service_orders"

    order_number: Mapped[str] = mapped_column(String, unique=True, nullable=False)
    client_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("clients.id"), index=True, nullable=False
    )
    vehicle_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("vehicles.id"), index=True, nullable=False
    )
    mechanic_id: Mapped[uuid.UUID | None] = mapped_column(
        Uuid, ForeignKey("mechanics.id"), index=True
    )
    reported_problem: Mapped[str | None] = mapped_column(String)
    diagnosis: Mapped[str | None] = mapped_column(String)
    status: Mapped[ServiceOrderStatus] = mapped_column(
        Enum(ServiceOrderStatus, name="service_order_status"),
        default=ServiceOrderStatus.CRIADA,
        server_default=ServiceOrderStatus.CRIADA.value,
        index=True,
        nullable=False,
    )
    total_amount: Mapped[Decimal] = mapped_column(
        Numeric(12, 2), default=Decimal("0"), server_default=text("0"), nullable=False
    )
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    client: Mapped[Client] = relationship(lazy="joined")
    vehicle: Mapped[Vehicle] = relationship(lazy="joined")
    mechanic: Mapped[Mechanic | None] = relationship(lazy="joined")
